Add basic support for .wh..wh..opq

tonistiigi opened the pull request:
This fixes the case where directory is removed in
aufs and then the same layer is imported to a
different graphdriver.

Currently when you do `rm -rf /foo && mkdir /foo`
in a layer in aufs the files under `foo` would
only be be hidden on aufs.

The problems with this fix:

1) When a new diff is recreated from non-aufs driver
the `opq` files would not be there. This should not
mean layer differences for the user but still
different content in the tar (one would have one
`opq` file, the others would have `.wh.*` for every
file inside that folder). This difference also only
happens if the tar-split file isn’t stored for the
layer.

2) New files that have the filenames before `.wh.wh..opq`
when they are sorted do not get picked up by non-aufs
graphdrivers. Fixing this would require a bigger
refactoring that is planned in the future.

<del>Note the `ExcludePatterns` change only affects the
files on the root so its not actually very essential
and could be removed to avoid the other `.wh*` garbage.</del>

Signed-off-by: Tonis Tiigi tonistiigi@gmail.com

replaces #15996

cc @jlhawn @icecrime

tonistiigi committed:
Add basic support for .wh..wh..opq

This fixes the case where directory is removed in
aufs and then the same layer is imported to a
different graphdriver.

Currently when you do `rm -rf /foo && mkdir /foo`
in a layer in aufs the files under `foo` would
only be be hidden on aufs.

The problems with this fix:

1) When a new diff is recreated from non-aufs driver
the `opq` files would not be there. This should not
mean layer differences for the user but still
different content in the tar (one would have one
`opq` file, the others would have `.wh.*` for every
file inside that folder). This difference also only
happens if the tar-split file isn’t stored for the
layer.

2) New files that have the filenames before `.wh..wh..opq`
when they are sorted do not get picked up by non-aufs
graphdrivers. Fixing this would require a bigger
refactoring that is planned in the future.


Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi commented:
I'm even more certain that we should keep the exclude check or fine tune it to only leave in `.opq`. If we don't exclude `.wh..wh.plnk` files the images can't be pushed to new repo after pull (https://github.com/docker/docker/issues/15643#issuecomment-143899374).

calavera commented:
I think leaving out of the image only opaque directories makes sense. I don't know if we can add the whiteout base directory out of the image too: `.wh..wh.aufs`.

tonistiigi commented:
I updated with exclusion patterns to exclude everything but `.opq`. Also added a `mkdir` to make sure that at least the empty directory exists if there are no more files.

calavera commented on pkg/archive/diff.go:
```diff
@@ -116,9 +119,23 @@ func UnpackLayer(dest string, layer Reader) (size int64, err error) {
-			if err := os.RemoveAll(originalPath); err != nil {
-				return 0, err
+			dir := filepath.Dir(path)
+			if originalBase == ".wh..opq" {
```
doesn't this name need an extra `.wh..`? Can we extract this to constants, so people know what they are?

tonistiigi replied:
Updated.

calavera replied:
very nice, thank you!

tonistiigi committed:
Add constants for AUFS whiteout files

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

calavera commented:
LGTM

jfrazelle commented:
LGTM
